src/pip/_internal/commands/install.py

```diff
@@ -341,6 +352,17 @@ def run(self, options: Values, args: List[str]) -> int:
                 reqs, check_supported_wheels=not options.target_dir
             )
 
+            if options.dry_run:
+                items = [
+                    f"{item.name}-{item.metadata['version']}"
+                    for item in sorted(
+                        requirement_set.all_requirements, key=lambda x: str(x.name)
+                    )
+                ]
+                if items:
+                    write_output("Would install %s", " ".join(items))
```

Comment:
I’m not entirely sure, but some requirements may have its `name` be None in the legacy resolver. It’s probably more reliable to use `metadata['name']`.

I would probably implement this like

```suggestion
                items = sorted(
                    r.metadata['name'], r.metadata['version']
                    for r in requirement_set.all_requirements
                )
                if items:
                    write_output(
                        "Would install %s",
                        " ".join("-".join(item) for item in items),
                    )
```